👤🔊 OP Accessibility: Screen Reader Improvements

sebastianbarry opened the pull request:
Part of issue: https://github.com/e-mission/e-mission-docs/issues/1011
This PR is for the purpose of taking over for @niccolopaganini, https://github.com/e-mission/e-mission-phone/pull/1112

Goals:

- Screen reader reads the entire page when Popups display, instead of only what's on the popup
- Unpronounceable buttons in accessibility inspector
- Profile buttons; some buttons screen-read as doubled the word
- Having screen reader read dropdowns as "dropdown" instead of "button"

sebastianbarry commented:
I also found this article that might be helpful for figuring out the popup - screen reader reading what's behind the popup issue: https://github.com/callstack/react-native-paper/issues/3166

sebastianbarry commented:
This is the i18n translation files PR for removing non-character characters https://github.com/e-mission/e-mission-translate/pull/53

sebastianbarry commented:
### I have completed most of the changes I want to for now. Here is a summary:

| # | Change | Description | Image | Tested on physical test device? |
|---|---|---|---|---|
| 1 | Logout button acts as 1 button | Before, the screenreader treated the logout button as a separate text element, and a icon button. Now, they should be paired as one and only one focusable component. |  | Not tested |
| 2 | Unpronounceable metrics card icons on Dashboard | Before, the metrics cards had icons that did not have accessibility text, meaning a screenreader user would not know what these icons represent. Now, their accessibility text is the same as the name of the icon ("details, graph" "user, aggregate"). |  | Not tested |
| 3 | Remove non-character characters, like 👇 | For the sake of accessibility users, non-character characters are considered bad form. Consider screen reader users, where an icon can either read as "unpronounceable", or, at best when labeled, can obscure the meaning of the icon (consider a text box that displays "⭐️ favorites" and screen reads "star favorites"; a screen reader user may think "what is a star favorite, and is it different than non-star favorites??"). Another example of where this can be an issue is in dyslexic-font users, whose fonts may not support non-character characters and they would display as empty boxes []. Any emojis have been removed, as well as in the translated files: https://github.com/e-mission/e-mission-translate/pull/53 |  | Not tested |
| 4 | BottomNavigation buttons read double: "Profile Profile" | Before, the screenreader read the BottomNavigation buttons as their text but doubled. I've added an accessibilityLabel field to the BottomNavigation elements, so now ideally the text will not be duplicated as it will be overridden by the accessibilityLabel property. The reason it reads twice is because React Native Paper displays 2 of the exact same text element, and one of them switches from 0 to 1 opacity, depending on whether it's selected or not. |  | Not tested |
| 5 | Remove focusable for icons in profile settings screen | Before, the screenreader would allow a user to focus the text and the icon of a SettingsRow element separately. On my personal phone, you can ONLY either select the text ("view privacy policy"), or the icon ("unpronounceable"). Now, they should not be focusable, and only the text or the entire SettingsRow as an accessibility element |  | Not tested |

sebastianbarry commented:
@JGreenlee I added this to your review board, if either you or @jiji14 have time I'd really appreciate you taking a look and providing feedback! :)

Next step is to test on a physical device. My test phone does not have the software update to support the current staging release of OpenPATH, so I need to work with @jiji14 to use theirs. However, I am not sure what's the process of doing this without pushing it to staging. @Abby-Wheelis or @JGreenlee do you know the process for getting the app to run with our custom code and test travel data, so that Jiji and I can test what the screen reader actually reads?

JGreenlee commented:
I only know how to do it for Android, but you can either:
1) install the devapp APK on a physical Android device and connect to the devapp server via IP address (the phone must be on the same Wi-Fi as the machine running the devapp server)
2) build the app with cordova, locate the output apk and install it on the Android phone (either by using ADB or manually copying the apk)

jiji14 commented:
@sebastianbarry 

I will test with an Android phone and comment here after I am done testing. I love how your summary is organized!

JGreenlee commented on www/i18n/en.json:
```diff
@@ -128,13 +128,13 @@
-    "select-mode-scroll": "Mode (👇 for more)",
-    "select-replaced-mode-scroll": "Replaces (👇 for more)",
-    "select-purpose-scroll": "Purpose (👇 for more)",
+    "choose-mode": "Mode",
+    "choose-replaced-mode": "Replaces",
+    "choose-purpose": "Purpose",
+    "choose-survey": "Add Trip Details",
+    "select-mode-scroll": "Mode (scroll for more)",
+    "select-replaced-mode-scroll": "Replaces (scroll for more)",
+    "select-purpose-scroll": "Purpose (scroll for more)",
```
We should make these changes for other languages too, right?

sebastianbarry replied:
I created a PR for Spanish and Lao! https://github.com/e-mission/e-mission-translate/pull/53

JGreenlee commented on www/i18n/en.json:
```diff
@@ -380,7 +381,8 @@
     "background-restrictions": "On Samsung and Huwaei phones, make sure that background restrictions are turned off",
-    "close": "Close"
+    "close": "Close",
+    "open-access-study": "Link to Open Access Study"
```
You can remove this - such a link is not planned at this time.

sebastianbarry replied:
Removed in https://github.com/e-mission/e-mission-phone/pull/1118/commits/bd9b2f2bd55c94c16bc4416db28529b435e2c028

JGreenlee commented on www/js/control/ExpandMenu.jsx:
```diff
@@ -16,6 +16,8 @@ const ExpansionSection = (props) => {
       title={t(props.sectionTitle)}
       titleStyle={rowStyles.title}
+      accessible={true}
+      accessibilityLabel={t(props.sectionTitle) + ' spinbutton'}
```
"spinbutton"? Are users going to know what that means?

sebastianbarry replied:
I used the React accessibilityRole documentation to come up with this (I hadn't heard of "spinbutton" either before reading this). It says:

> - **spinbutton** used to represent a button that opens a list of choices.

There are no other options I saw in this list that really matched the criteria better. I briefly spoke to a friend of mine who uses accessibility features and they said that most people they know who use screen reader have a pretty good idea of how web elements work purely by using the screenreader, so I wouldn't be surprised if spinbutton is a word that is commonly known among people who use screenreader. I'd be open to changing this if we find a better word or role to use.

https://reactnative.dev/docs/accessibility#accessibilityrole

shankari replied:
We can tweak further based on feedback from TCAT

JGreenlee commented on www/js/components/ToggleSwitch.tsx:
```diff
@@ -11,6 +11,7 @@ const ToggleSwitch = ({ value, buttons, ...rest }: SegmentedButtonsProps) => {
       buttons={buttons.map((o) => ({
         value: o.value,
         icon: o.icon,
+        accessibilityLabel: o.value,
```
Side-note that `o.value` here is not really meant to be user-facing text; it's moreso a string used as key for the different toggle options. (e.g. `"details"` /`"graph"`, `"user"` /`"aggregate"`, `"labeled"` / `"detected"`)

I know we don't have translations for those things right now so this is better than nothing, but ideally we'd have dedicated labels

JGreenlee commented on www/i18n/en.json:
```diff
@@ -128,13 +128,13 @@
-    "select-mode-scroll": "Mode (👇 for more)",
-    "select-replaced-mode-scroll": "Replaces (👇 for more)",
-    "select-purpose-scroll": "Purpose (👇 for more)",
+    "choose-mode": "Mode",
+    "choose-replaced-mode": "Replaces",
+    "choose-purpose": "Purpose",
```
Is there some way we can keep the emojis and just silence them to screen readers? I think at least some of these are valuable as a visual aid to sighted users, but I understand that they could make it more confusing for those using screen readers.
It seems like we should be able to keep emojis that are more 'decorative' in nature.

sebastianbarry replied:
As a sighted user, I agree that these emojis help with clarity and look better overall.

My consideration in removing these emojis was less to avoid screen reader "unpronounceable" bugs, and more to avoid the use of non-character characters in a general sense for broader accessibility. Best case scenario, the screen reader skips over these types of characters entirely, but then we might see users with a dyslexic font app having empty boxes [] for these. I did not directly test this, so I think this is a great discussion. I talked more about my thoughts in #3's description:

>

shankari replied:
I think this is fine for now, but may be a question for Anat and Kunal as they re-start testing the app.

JGreenlee commented on www/i18n/en.json:
```diff
@@ -367,6 +367,7 @@
     "welcome-to-app": "Welcome to <b>{{appName}}!</b>",
     "app-name": "NREL OpenPATH",
     "to-proceed-further": "To proceed further, please scan or paste an access code that has been provided by your program administrator through a website, email, text message, or printout.",
+    "op-acc-link": "Link to Open-Access study",
```
You can remove this - such a link is not planned at this time.

sebastianbarry replied:
```suggestion
```

JGreenlee reviewed (commented):
Thanks for taking on this work and addressing a good number of these issues.
This review is not conclusive and we still have testing to do -- but I do want to leave an initial batch of comments here

JGreenlee commented on package-lock.json:
remove this file please

sebastianbarry replied:
Deleted https://github.com/e-mission/e-mission-phone/pull/1118/commits/d5b7b51fe20fe871b30122d027efd4a58be9c583

sebastianbarry commented on www/i18n/en.json:
```diff
@@ -380,7 +380,8 @@
     "dont-force-kill": "Do not force kill the app",
     "background-restrictions": "On Samsung and Huwaei phones, make sure that background restrictions are turned off",
-    "close": "Close"
+    "close": "Close",
+    "open-access-study": "Link to Open Access Study"
```
```suggestion
    "close": "Close"
```

jiji14 commented:
@sebastianbarry Certainly! Here are the test results! I recorded them for your reference. Please let me know if you have any questions!

## Test Device Information
Android - Samsung Galaxy S20

## 1. Logout button acts as 1 button 
### FAIL
`Logout` text is still separated

https://github.com/e-mission/e-mission-phone/assets/47590587/28b8fbb4-de2d-46ba-bfdf-a96cb453ff0d


## 2. Unpronounceable metrics card icons on Dashboard
### PASS

https://github.com/e-mission/e-mission-phone/assets/47590587/1ac6718a-58e4-4e85-bf36-87340f983ef4


## 3. Remove non-character characters, like 👇
### PASS

https://github.com/e-mission/e-mission-phone/assets/47590587/f63977b8-c28c-40f2-9cd0-8c5efc71d567


## 4. BottomNavigation buttons read double: "Profile Profile"
### FAIL
Tab buttons still read double ex) `Label tab Label Label` 

https://github.com/e-mission/e-mission-phone/assets/47590587/e77bfc06-f776-4764-82c5-071a66662779


## 5. Remove focusable for icons in profile settings screen
### PASS

https://github.com/e-mission/e-mission-phone/assets/47590587/1d32036c-7a40-49d5-b652-71a27b591fb1


## 6. Read dropdown as "spinbutton" in profile settings screen
### PASS
It says `Developer zone spinbutton button`. I think it would be better if it just says `Developer zone spinbutton` but at least it says `spinbutton`!

https://github.com/e-mission/e-mission-phone/assets/47590587/361adb21-eef9-45a2-ab9c-ebe968312dc9

sebastianbarry commented:
@jiji14 Great testing!! I really appreciate your detail and videos, very helpful!

### For the failed tests...:

## Test 1: Logout button acts as 1 button

I fixed this in https://github.com/e-mission/e-mission-phone/pull/1118/commits/b94abff5991c0bea567951fa4c7558cc915233b5 . It should now display as one single button, like the settings row buttons do. I modeled it after them, by changing the type of element that the logout button is.

<img width="683" alt="Screenshot 2024-01-22 at 10 28 05 AM" src="https://github.com/e-mission/e-mission-phone/assets/61334340/a85c3664-c196-4841-9995-b286e52d3356">

## Test 4: BottomNavigation buttons read double: "Profile Profile"

This issue may be out of our control. I haven't changed anything, although when I test with my iPhone emulator, it does not read "Label tab label label... etc." It only reads "Label tab, Dashbaord tab, profile tab". This makes me think that what the screen reader reads is dependent upon the device's built in screen reader, so it is different from iOS to Android. I think the solution will require creating a pull request to `react-native-paper`'s codebase. **@JGreenlee Is this something we want to pursue?**

https://github.com/e-mission/e-mission-phone/assets/61334340/7e9d3a39-7bab-4d92-8b16-67c545d421e0

JGreenlee commented:
@sebastianbarry Can you toggle the target branch to `master` and back to `service_rewrite_2023`?
I want to verify that I resolved merge conflicts correctly

sebastianbarry commented:
@JGreenlee Did it! let me know if it worked :)

sebastianbarry commented:
@JGreenlee I rebased it back to `master`! Looks like the only commit you made on this PR now is that one merge commit. There were 3 merge conflicts when I moved it over, but I took care of those since they were very simple.

JGreenlee commented:
Glad that resolved itself! However, the 'build' test is now failing.
I haven't seen that happen before.

JGreenlee commented:
I just tried testing this branch in the devapp on an Android phone with a screenreader. TalkBack did not read anything at all and I couldn't navigate to anything by swiping left/right.

Confused, I tried TalkBack on the latest NREL OpenPATH. I got the same result.
I tried the current `master` on the devapp, and I also tried `master` at the point when the previous feature branch was merged in 342e1908d4d1dbbcc15610e8492218d769b52c74.

On all of these, TalkBack was not reading anything. Talkback works for other apps on the phone.

It didn't used to be like this, so either we had a major accessibility regression at some point (at least for Android), or I'm not remembering correctly how I did it before.

JGreenlee commented:
When I get time, I will go back even further and see if I can either isolate the regression, or determine that there is no regression

sebastianbarry commented:
At least for my iOS simulator, the accessibility inspector (simulating VoiceOver) is working on `screen_reader` branch

![Screenshot 2024-02-06 at 12 19 24 PM](https://github.com/e-mission/e-mission-phone/assets/61334340/96458c38-9638-4f63-b857-907bd52ba328)

As for the build fail, I'm not sure what that is. I'm not even quite sure what the test fail error message is; is it this?

```
npm ERR! code 128
npm ERR! An unknown git error occurred
npm ERR! command git --no-replace-objects ls-remote ssh://git@github.com/e-mission/cordova-unified-logger.git
npm ERR! git@github.com: Permission denied (publickey).
npm ERR! fatal: Could not read from remote repository.
npm ERR! 
npm ERR! Please make sure you have the correct access rights
npm ERR! and the repository exists.
```

JGreenlee reviewed (approved):
I haven't been able to test this out on a physical device, but Jiji and Sebastian both tested. The code changes look good to me so I will push it through to 'ready for review'

shankari commented:
The current error is

```
Failed to restore plugin "cordova-plugin-em-unifiedlogger". You might need to try adding it again. Error: CordovaError: Failed to fetch plugin git+https://github.com/e-mission/cordova-unified-logger.git#v1.3.6 via registry.
Probably this is either a connection problem, or plugin spec is incorrect.
Check your connection and plugin name/version/URL.
CordovaError: Error: npm: Command failed with exit code 128 Error output:
npm ERR! code 128
npm ERR! An unknown git error occurred
npm ERR! command git --no-replace-objects ls-remote ssh://git@github.com/e-mission/cordova-unified-logger.git
npm ERR! git@github.com: Permission denied (publickey).
npm ERR! fatal: Could not read from remote repository.
npm ERR! 
npm ERR! Please make sure you have the correct access rights
npm ERR! and the repository exists.
```

Re-running to check if that resolves the error and passes the tests

sebastianbarry commented:
@JGreenlee @shankari it passed! Thank you for your help

shankari commented:
To summrarize the current state:
- it is not clear if this even works on android, however, this may not be a regression since it was not working even without these changes
- BottomNavigation buttons read double: "Profile Profile": is not fixed on android but is fixed on iOS. @sebastianbarry thinks that this will need changes to React Native Paper, but I don't see any investigation/evidence around this.

Regardless, it looks like it at least works on iOS, and is a step forward, so I will plan on reviewing + merging. @sebastianbarry  can you file a follow up issue so that we can finish polishing this later?

shankari commented on www/js/control/SettingRow.tsx:
```diff
@@ -12,18 +12,24 @@ type Props = {
   let rightComponent;
   if (iconName) {
-    rightComponent = <List.Icon icon={iconName} />;
+    rightComponent = <List.Icon icon={iconName} aria-hidden={true} />;
```
I wonder if manually setting the accessibilityLabel to "" here would fix the double repetition issue. I would also suggest looking at the react native paper code (it is open-source!) to see if this has been reported by others and whether there are any workarounds.

sebastianbarry replied:
Lets figure this out in the subsequent PR. New issue created: https://github.com/e-mission/e-mission-docs/issues/1049

shankari commented on www/js/control/SettingRow.tsx:
```diff
@@ -12,18 +12,24 @@ type Props = {
+        value={switchValue}
+        accessibilityLabel={t(textKey as any)}
+        accessibilityHint={switchValue ? t('Currently enabled') : t('Currently disabled')}
+      />
+    );
   }
   let descriptionText;
   if (desc) {
-    descriptionText = { desc };
+    descriptionText = desc;
   } else {
```
I don't understand the reason for this

sebastianbarry replied:
Resolved in https://github.com/e-mission/e-mission-phone/pull/1118/commits/b0596e0741eee36a946b68115468b62aef7671c8

shankari commented on www/js/control/SettingRow.tsx:
```diff
@@ -33,9 +39,12 @@ const SettingRow = ({ textKey, iconName, action, desc, switchValue, descStyle }:
       style={styles.item(colors.surface)}
       title={t(textKey as any)}
       titleStyle={styles.title}
-      description={desc}
+      description={descriptionText}
```
and this. The related commit is: https://github.com/sebastianbarry/e-mission-phone/commit/69e3bee47e0aa4f4623f551ab7e14721182be8dd which does not have any explanation.

sebastianbarry replied:
Resolved in https://github.com/e-mission/e-mission-phone/pull/1118/commits/3090f33606c18a92af570483e0e7a837ba2086f2

shankari commented on www/js/onboarding/WelcomePage.tsx:
```diff
@@ -100,6 +100,7 @@ const WelcomePage = () => {
       <Surface style={[onboardingStyles.page, { paddingVertical: 0 }]}>
         <View style={s.headerArea(windowWidth, colors)} aria-hidden={true} />
         <IconButton
+          accessibilityLabel="More info"
```
this is not internationalized

sebastianbarry replied:
Resolved in https://github.com/e-mission/e-mission-phone/pull/1118/commits/cf7c86048930c54409a60e5eab09582ef4332cfc

sebastianbarry replied:
Associated commit https://github.com/e-mission/e-mission-translate/pull/53/commits/361fa6785d85050be3ecb04c649ff65baad81511 in PR https://github.com/e-mission/e-mission-translate/pull/53

shankari reviewed (changes requested):
@sebastianbarry just a few comments to fix. The main change that I would really like to include is the i18n

sebastianbarry commented on www/js/control/SettingRow.tsx:
```diff
@@ -33,9 +39,12 @@ const SettingRow = ({ textKey, iconName, action, desc, switchValue, descStyle }:
       title={t(textKey as any)}
       titleStyle={styles.title}
-      description={desc}
+      description={descriptionText}
```
```suggestion
      description={desc}
```

sebastianbarry commented on www/js/control/SettingRow.tsx:
```diff
@@ -12,18 +12,24 @@ type Props = {
   let descriptionText;
   if (desc) {
-    descriptionText = { desc };
+    descriptionText = desc;
```
```suggestion
    descriptionText = { desc };
```

sebastianbarry reviewed (commented):
Undoing `desc` and `descriptionText` changes in www/js/control/SettingRow.tsx

These are old changes from before I took over the PR, I can go ahead undo them

sebastianbarry commented:
@shankari I addressed your comments. I think once we pass the tests, we should be good to merge!!
